File: integration_tests/dart_client_tests/test/upload_test.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_uploads.dart';
import 'fixtures.dart';
import 'mock_adapter.dart';

void main() {
  late MockAdapter adapter;
  late MockMultipartAdapter multipart;
  late GraphLinkClient client;

  GLUpload makeFile(List<int> bytes, {String? filename}) => GLUpload.fromBytes(
        Uint8List.fromList(bytes),
        filename: filename,
        mimeType: 'image/png',
      );

  setUp(() {
    adapter = MockAdapter();
    multipart = MockMultipartAdapter();
    client = GraphLinkClient(
      adapter: adapter.call,
      uploadAdapter: multipart.call,
      wsAdapter: MockWebSocketAdapter(),
    );
  });

  // ── uploadOneFile ─────────────────────────────────────────────────────────

  group('uploadOneFile — single file', () {
    setUp(() {
      multipart.registerData('uploadOneFile', {'uploadOneFile': kUserAliceJson});
    });

    test('response deserialized: id correct', () async {
      final res = await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      expect(res.uploadOneFile.id, equals('user-1'));
    });

    test('response deserialized: name correct', () async {
      final res = await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      expect(res.uploadOneFile.name, equals('Alice Smith'));
    });

    test('multipart adapter is called once', () async {
      await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      expect(multipart.callCount, equals(1));
    });

    test('operations contains correct userId variable', () async {
      await client.mutations
          .uploadOneFile(userId: 'user-42', file: makeFile([1, 2, 3]));
      final ops = jsonDecode(multipart.lastCall!['operations'] as String)
          as Map<String, dynamic>;
      expect(ops['variables']['userId'], equals('user-42'));
    });

    test('file variable is null in operations (multipart spec)', () async {
      await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      final ops = jsonDecode(multipart.lastCall!['operations'] as String)
          as Map<String, dynamic>;
      expect(ops['variables']['file'], isNull);
    });

    test('map slot 0 points to variables.file', () async {
      await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      final map = jsonDecode(multipart.lastCall!['map'] as String)
          as Map<String, dynamic>;
      expect(map['0'], equals(['variables.file']));
    });

    test('file part is present under slot key "0"', () async {
      await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      expect(multipart.lastCall!.containsKey('0'), isTrue);
    });

    test('regular JSON adapter is not called', () async {
      await client.mutations
          .uploadOneFile(userId: 'user-1', file: makeFile([1, 2, 3]));
      expect(adapter.callCount, equals(0));
    });
  });

  // ── uploadFileList ────────────────────────────────────────────────────────

  group('uploadFileList — list of files', () {
    setUp(() {
      multipart.registerData(
          'uploadFileList', {'uploadFileList': kUserAliceJson});
    });

    test('response deserialized: id correct', () async {
      final res = await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
      );
      expect(res.uploadFileList.id, equals('user-1'));
    });

    test('files variables are all null in operations (multipart spec)', () async {
      await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
      );
      final ops = jsonDecode(multipart.lastCall!['operations'] as String)
          as Map<String, dynamic>;
      expect(ops['variables']['files'], equals([null, null]));
    });

    test('map slot 0 points to variables.files.0', () async {
      await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
      );
      final map = jsonDecode(multipart.lastCall!['map'] as String)
          as Map<String, dynamic>;
      expect(map['0'], equals(['variables.files.0']));
    });

    test('map slot 1 points to variables.files.1', () async {
      await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
      );
      final map = jsonDecode(multipart.lastCall!['map'] as String)
          as Map<String, dynamic>;
      expect(map['1'], equals(['variables.files.1']));
    });

    test('both file parts present in parts map', () async {
      await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
      );
      expect(multipart.lastCall!.containsKey('0'), isTrue);
      expect(multipart.lastCall!.containsKey('1'), isTrue);
    });

    test('single-file list produces one slot', () async {
      await client.mutations.uploadFileList(
        userId: 'user-1',
        files: [makeFile([1, 2, 3])],
      );
      final map = jsonDecode(multipart.lastCall!['map'] as String)
          as Map<String, dynamic>;
      expect(map.length, equals(1));
    });
  });
}

File: integration_tests/typescript_client_tests/test/upload.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import type { GLUpload } from '../lib/generated/client/graph-link-uploads.js';
import { MockAdapter, MockMultipartAdapter, MockWsAdapter } from './mock-adapter.ts';
import { kUserAliceJson } from './fixtures.ts';

function makeFile(bytes: number[], filename = 'test.png'): GLUpload {
  return {
    stream: new Blob([new Uint8Array(bytes)]),
    length: bytes.length,
    filename,
    mimeType: 'image/png',
  };
}

let adapter: MockAdapter;
let multipart: MockMultipartAdapter;
let client: GraphLinkClient;

beforeEach(() => {
  adapter = new MockAdapter();
  multipart = new MockMultipartAdapter();
  client = new GraphLinkClient(adapter.call, new MockWsAdapter(), multipart.call);
});

// ── uploadOneFile ─────────────────────────────────────────────────────────────

describe('uploadOneFile — single file', () => {
  beforeEach(() => {
    multipart.registerData('uploadOneFile', { uploadOneFile: kUserAliceJson });
  });

  it('response deserialized: id correct', async () => {
    const res = await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    expect(res.uploadOneFile.id).toBe('user-1');
  });

  it('response deserialized: name correct', async () => {
    const res = await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    expect(res.uploadOneFile.name).toBe('Alice Smith');
  });

  it('multipart adapter is called once', async () => {
    await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    expect(multipart.callCount).toBe(1);
  });

  it('userId variable is in operations payload', async () => {
    await client.mutations.uploadOneFile({ userId: 'user-42', file: makeFile([1, 2, 3]) });
    const ops = JSON.parse(multipart.lastCall!.parts['operations'] as string);
    expect(ops.variables.userId).toBe('user-42');
  });

  it('file variable is null in operations (multipart spec)', async () => {
    await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    const ops = JSON.parse(multipart.lastCall!.parts['operations'] as string);
    expect(ops.variables.file).toBeNull();
  });

  it('map slot 0 points to variables.file', async () => {
    await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    const map = JSON.parse(multipart.lastCall!.parts['map'] as string);
    expect(map['0']).toEqual(['variables.file']);
  });

  it('file part is present under slot key "0"', async () => {
    await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    expect(multipart.lastCall!.parts['0']).toBeDefined();
  });

  it('regular JSON adapter is not called', async () => {
    await client.mutations.uploadOneFile({ userId: 'user-1', file: makeFile([1, 2, 3]) });
    expect(adapter.callCount).toBe(0);
  });
});

// ── uploadFileList ────────────────────────────────────────────────────────────

describe('uploadFileList — list of files', () => {
  beforeEach(() => {
    multipart.registerData('uploadFileList', { uploadFileList: kUserAliceJson });
  });

  it('response deserialized: id correct', async () => {
    const res = await client.mutations.uploadFileList({
      userId: 'user-1',
      files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
    });
    expect(res.uploadFileList.id).toBe('user-1');
  });

  it('files variables are all null in operations (multipart spec)', async () => {
    await client.mutations.uploadFileList({
      userId: 'user-1',
      files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
    });
    const ops = JSON.parse(multipart.lastCall!.parts['operations'] as string);
    expect(ops.variables.files).toEqual([null, null]);
  });

  it('map slot 0 points to variables.files.0', async () => {
    await client.mutations.uploadFileList({
      userId: 'user-1',
      files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
    });
    const map = JSON.parse(multipart.lastCall!.parts['map'] as string);
    expect(map['0']).toEqual(['variables.files.0']);
  });

  it('map slot 1 points to variables.files.1', async () => {
    await client.mutations.uploadFileList({
      userId: 'user-1',
      files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
    });
    const map = JSON.parse(multipart.lastCall!.parts['map'] as string);
    expect(map['1']).toEqual(['variables.files.1']);
  });

  it('both file parts present in parts map', async () => {
    await client.mutations.uploadFileList({
      userId: 'user-1',
      files: [makeFile([1, 2, 3]), makeFile([4, 5, 6])],
    });
    expect(multipart.lastCall!.parts['0']).toBeDefined();
    expect(multipart.lastCall!.parts['1']).toBeDefined();
  });

  it('single-file list produces exactly one slot in map', async () => {
    await client.mutations.uploadFileList({ userId: 'user-1', files: [makeFile([1, 2, 3])] });
    const map = JSON.parse(multipart.lastCall!.parts['map'] as string);
    expect(Object.keys(map)).toHaveLength(1);
  });
});
